### ISSUE TITLE
Question about incompatibilities created at version picking

### ISSUE THREAD

**Issue by mpizenberg** [dart-lang/pub#2485]
Hi, I'm starting to work on error reporting for an implementation of PubGrub in Elm and something is not clear to me regarding incompatibilities created if version picking of a selected package during decision making failed. I have two questions, mainly.

(1) Is an incompatibility created during version picking considered "derived" or "external". I'd describe it as derived since it is true only due to our configuration but I'm not sure how to pick its causes.

Concretely, below is my current implementation. It takes a partial solution, a function providing all available versions of a package, and returns either `Nothing` or `Just (Ok (package, version))` or `Just (Err (package, incompatibility))` where the incompability is built as `{term}`, as described in decision making (https://github.com/dart-lang/pub/blob/master/doc/solver.md#decision-making).

```elm
pickPackageVersion : PartialSolution -> (String -> List Version) -> Maybe (Result ( String, Incompatibility ) ( String, Version ))
pickPackageVersion partial listAvailableVersions =
    case pickPackage partial of
        Just ( name, term ) ->
            pickVersion (listAvailableVersions name) term
                |> Maybe.map (Tuple.pair name)
                |> Result.fromMaybe ( name, Incompatibility.fromTerm name term )
                |> Just

        Nothing ->
            Nothing
```

From my understanding, all incompatibilities should be true at all times. However an incompatibility such as this one only makes sense with regard to the current state of the partial solution, that may be backtracked later? So my second question (2) is: Isn't this an issue in the algorithm or is there something else I'm not following?

**Comment by mpizenberg** [comment]
In my debug log, when running the [example 5 (linear error reporting)](https://github.com/dart-lang/pub/blob/master/doc/solver.md#linear-error-reporting), I get the following.
```
Final incompatibility:
   root: 1.0.0  <<<  derived from:
      root: 1.0.0, foo: 1.0.0 <= v < 2.0.0  <<<  derived from:
         root: 1.0.0, baz: Not ( 1.0.0 <= v < 2.0.0 )  <<<  from dependency of root at version 1.0.0
         foo: 1.0.0 <= v < 2.0.0, baz: Not ( 3.0.0 <= v < 4.0.0 )  <<<  derived from:
            bar: 2.0.0 <= v < 3.0.0, baz: Not ( 3.0.0 <= v < 4.0.0 )  <<<  derived from:
               # bar: 2.0.0, baz: Not ( 3.0.0 <= v < 4.0.0 )  <<<  from dependency of bar at version 2.0.0
               # bar: 2.0.1 <= v < 3.0.0  <<<  from unknown reason ...
            foo: 1.0.0 <= v < 2.0.0, bar: Not ( 2.0.0 <= v < 3.0.0 )  <<<  derived from:
               # foo: 1.0.0, bar: Not ( 2.0.0 <= v < 3.0.0 )  <<<  from dependency of foo at version 1.0.0
               # foo: 1.0.1 <= v < 2.0.0  <<<  from unknown reason ...
      root: 1.0.0, foo: Not ( 1.0.0 <= v < 2.0.0 )  <<<  from dependency of root at version 1.0.0: ""
```
But in the explanation documentation of example 5, the deepest levels of my derivation tree, marked here with `#` are not present in the derivation graph. Is it always the case that those can be removed and their parent of the form `first any, second not ...`?

**Comment by sigurdm** [comment]
> (1) Is an incompatibility created during version picking considered "derived" or "external". I'd describe it as derived since it is true only due to our configuration but I'm not sure how to pick its causes.

It is external.
The incompatibility is actually valid under all configurations. For example a dependency from `foo 1.0.0` to `bar 1.0.0` gives the following incompatibility: {`foo 1.0.0`, `not bar 1.0.0`}. 
That is always true that you cannot at the same time have `foo 1.0.0` but not have `bar 1.0.0`, but it is only added as an incompatibility because we are exploring foo 1.0.0. Therefore it doesn't have any causes.

In the pub solver these incompatibilities are added here: https://github.com/dart-lang/pub/blob/master/lib/src/solver/package_lister.dart#L280 and they indeed have no causes.

**Comment by mpizenberg** [comment]
> For example a dependency from foo 1.0.0 to bar 1.0.0 gives the following incompatibility: {foo 1.0.0, not bar 1.0.0}.

Yep that was already clear to me. I was referring not to an incompatibility created from a dependency but from failing to pick a version after having chose a package. The bit that corresponds to:

> If there is no such version, add an incompatibility {term} to the incompatibility set and return package's name. This tells Pubgrub to avoid this range of versions in the future.

**Comment by sigurdm** [comment]
Sorry I misunderstood you. And sorry for the long delay in response.

This kind of incompatibility is added here: https://github.com/dart-lang/pub/blob/master/lib/src/solver/version_solver.dart#L380
```
        _addIncompatibility(Incompatibility(
            [Term(package, true)], IncompatibilityCause.noVersions));
```
For a while I thought you where right. We cannot say that `package` is not acceptable just because it doesn't fit our current constraints. But `package` is here actually a package-range. And if we once have decided there exists no versions in that range, then we might as well remember that in an incompatibility.

Does that make more sense?

**Comment by mpizenberg** [comment]
No worry, that's a personal project so I'm in no hurry :)

I think I understand now! Let me know if I'm saying something wrong in the following.
What this means is that there simply exists no version in that range and the way we computed that range is totally independent of that fact. So we are just recording in an incompatibility the fact that such range does not contain anything, no need to ask the agent that has lazy access to versions about it later.

I suppose this enables filling gaps in incompatibilities. For example, after realizing that `A 1.0.0` is incompatible and looking for an A in the range 1.0.1 <= A < 2.0.0, if there is no version there, this will be recorded as { A >= 1.0.1 < 2.0.0 } and "smartly" merged with { A = 1.0.0 } into { A ^1.0.0 } with the same cause as the one used for { A = 1.0.0 }.

I had not implemented "smart" incompatibility merges yet and I suppose this is why I still have those in my derivation graph while they are not in the provided examples. Am I still correct?
